### PR TITLE
Fixe #4177: mockRuleStore require LoadRuleGroups, and fix ListAllRules

### LINKED ISSUE #4177: `/ruler/rule_groups` doesn't work with object stores
`/ruler/rule_groups` endpoint added in #3529 doesn't work when used with object store. Problem is that code calls [`ListRuleGroupsForUserAndNamespace`](https://github.com/cortexproject/cortex/blob/5d55a28c796dc1f44a864a0c701744f8f3f4e5cb/pkg/ruler/ruler.go#L863) to find all rule groups, but it doesn't call `LoadRuleGroups` required to actually *load* those rule groups. This step is necessary for some stores, eg. when using object store.

Tests currently pass, because `mockRuleStore` doesn't require this extra step.

Expected fix should change `mockRuleStore` to *require* `LoadRuleGroups`, and also fix `ListAllRules` to perform this load.

### PR DESCRIPTION
**Pull request by AllenzhLi**
Signed-off-by: allenzhli <allenzhli@tencent.com>

**What this PR does**:

**Which issue(s) this PR fixes**:
Fixes #4177

**Checklist**
- [x] Tests updated
- [x] `CHANGELOG.md` updated - the order of entries should be `[CHANGE]`, `[FEATURE]`, `[ENHANCEMENT]`, `[BUGFIX]`

### CODE REVIEW

File: pkg/ruler/store_mock_test.go [hunk 1]
```diff
@@ -131,6 +134,21 @@ func (m *mockRuleStore) ListAllRuleGroups(_ context.Context) (map[string]rulespb
 	m.mtx.Lock()
 	defer m.mtx.Unlock()
 
+	if m.loaded {
```
Comment (pstibrany) [inline pkg/ruler/store_mock_test.go:137]: Can we simply build the result on each call, without changing internal state of `mockRuleStore`? I find `m.loaded` confusing, and not sure what it is trying to express.

In other words, I would suggest to remove `m.loads` and `m.loaded`, and just build the result on each call to `ListAllRuleGroups` ("unloaded"). Same for `ListRuleGroupsForUserAndNamespace`. WDYT?

**Review by pstibrany — COMMENTED** [review]
Thank you for working on this.

File: pkg/ruler/store_mock_test.go (hunk 1 above)
Comment (AllenzhLi) [reply pkg/ruler/store_mock_test.go:137]: If we want change `mockRuleStore` to require LoadRuleGroups, then we should not return `Rules` in `ListAllRuleGroups`, and fill `Rules` in `LoadRuleGroups `. 

The problem is that in `mockRuleStore` we have no other storage to store that `Rules`.

Comment (pstibrany) [reply pkg/ruler/store_mock_test.go:137]: Idea was that `ListAllRuleGroups` would build the result, keeping only `Namespace`, `Name` and `User` fields. Similar to what your code already does, but without extra fields in `mockRuleStore`.

Comment (AllenzhLi) [reply pkg/ruler/store_mock_test.go:137]: Get it, thanks.

File: pkg/ruler/store_mock_test.go [hunk 2]
```diff
@@ -148,6 +156,14 @@ func (m *mockRuleStore) ListRuleGroupsForUserAndNamespace(_ context.Context, use
 		return rulespb.RuleGroupList{}, nil
 	}
 
+	for i, r := range userRules {
```
Comment (pstibrany) [inline pkg/ruler/store_mock_test.go:159]: I'm not sure how `ListRuleGroupsForUserAndNamespace` got this big, but can we simplify it to just:

```
func (m *mockRuleStore) ListRuleGroupsForUserAndNamespace(_ context.Context, userID, namespace string) (rulespb.RuleGroupList, error) {
	m.mtx.Lock()
	defer m.mtx.Unlock()

	var result rulespb.RuleGroupList
	for _, r := range m.rules[userID] {
		if namespace != "" && namespace != r.Namespace {
			continue
		}

		result = append(result, &rulespb.RuleGroupDesc{
			Namespace: r.Namespace,
			Name:      r.Name,
			User:      userID,
		})
	}
	return result, nil
}
```

**Review by pstibrany — APPROVED** [review]
Thank you, LGTM.

I've left one suggestion to simplify `ListRuleGroupsForUserAndNamespace`, but that's not a blocker.

File: pkg/ruler/store_mock_test.go (hunk 2 above)
Comment (AllenzhLi) [reply pkg/ruler/store_mock_test.go]: Wonderful! All is work well.

**Review by pracucci — APPROVED** [review]
LGTM!

**Pull request merged**